### DISCUSSION TITLE
micTObt ? mic2monitor and sinus2a2dp work fine

### DISCUSSION THREAD

**Discussion by monopol40** [pschatzmann/ESP32-A2DP#267] [Q&A] [answered]
/*

QUESTION: Hallo Herr pschatzmann, micTOmonitor in printBuffer and sinusTOa2dp in get_data_channels work fine. I want to change get_data_channels to work with mic instead of sinus signal, eg. micTOa2dp. I want to hear micro on BT Phone . I get stuck and tried several examples.  Any hints would be great! Thanks in advance!!

  
  Streaming of sound data with Bluetooth to other Bluetooth device.
  We generate 2 tones which will be sent to the 2 channels.
  
  Copyright (C) 2020 Phil Schatzmann
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/
#include "BluetoothA2DPSource.h"
#include <math.h> 

#include "Arduino.h"
#include "AudioTools.h"
#include "AudioKitHAL.h"

#include "AudioLibs/AudioKit.h"
using namespace audio_tools;

AudioKit kit; 

#define c3_frequency  130.81

BluetoothA2DPSource a2dp_source;

const int BUFFER_SIZE = 1024;
uint8_t buffer[BUFFER_SIZE];

// The supported audio codec in ESP32 A2DP is SBC. SBC audio stream is encoded
// from PCM data normally formatted as 44.1kHz sampling rate, two-channel 16-bit sample data
int32_t get_data_channels(Frame *frame, int32_t channel_len) {
    //Serial.println(channel_len);  //128
    static double m_time = 0.0;
    double m_amplitude = 10000.0;  // -32,768 to 32,767
    double m_deltaTime = 1.0 / 44100.0;
    double m_phase = 0.0;
    double double_Pi = PI * 2.0;
    // fill the channel data
    for (int sample = 0; sample < channel_len; ++sample) {
        double angle = double_Pi * c3_frequency * m_time + m_phase;
        frame[sample].channel1 = m_amplitude * sin(angle);
        frame[sample].channel2 = frame[sample].channel1;
        m_time += m_deltaTime;
    }
    return channel_len;
}

void printBuffer(int len){
  // by default we get int16_t values on 2 channels = 4 bytes per frame
  int16_t *value_ptr = (int16_t*) buffer;
  for (int j=0;j<len/4;j++){
    Serial.print(*value_ptr++);
    Serial.print(", ");
    Serial.println(*value_ptr++);
  }
}

void setup() {
  Serial.begin(115200);
  auto cfg = kit.defaultConfig(AudioInput);
  cfg.adc_input = AUDIO_HAL_ADC_INPUT_LINE1; // microphone
  cfg.sample_rate = AUDIO_HAL_16K_SAMPLES;
  kit.begin(cfg);
  a2dp_source.start("BT-OEH1", get_data_channels); 
} 

void loop() {
   // to prevent watchdog in release > 1.0.6
  delay(10);

  size_t len = kit.read(buffer, BUFFER_SIZE);
  printBuffer(len);
 }

**Comment by monopol40** [comment]
addendum: I'm using exp32 audio kit V2.2 with ac101 
Board Number  6) ai_thinker (AC101) 2762 2957

**Comment by pschatzmann** [comment]
I suggest that you have a look at the https://github.com/pschatzmann/arduino-audio-tools project.
The https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-basic-api/base-i2s-a2dp contains a corrsponding example. You can just replace the I2SStream with a AudioKitStream as input.

The https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit contains plenty of examples that show how to use the audiokit as input.

**Reply by monopol40** [reply]
Many thanks for your answer. I tried to setup your hint above replacing I2SStream with a AudiKitStream as input in get_sound_data, changing i2s.readBytes with kit.readBytes. But this does not work. 
Herr pschatzmann, höflichst frage ich, ist vielleicht const size_t max_buffer_len = 150; zu ändern? Oder gibts noch eine andere Idee?
Vielen Dank im Voraus!

Hier der Sketch:

/**
 * @file base-i2s-a2dp.ino
 * @author Phil Schatzmann
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/base-i2s-a2dp/README.md
  * @copyright GPLv3
*/

#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"

#include "AudioLibs/AudioKit.h"

/**
 * @brief We use a INMP441 I2S microphone as input and send the data to A2DP
 * Unfortunatly the data type from the microphone (int32_t)  does not match with the required data type by A2DP (int16_t),
 * so we need to convert.
 * 
 */ 

BluetoothA2DPSource a2dp_source;
//I2SStream i2s;
AudioKitStream kit;
ConverterFillLeftAndRight<int16_t> bothChannels(LeftIsEmpty);
const size_t max_buffer_len = 150;
const int channels = 2;
const size_t max_buffer_bytes = max_buffer_len * sizeof(int16_t) * channels;
uint8_t buffer[max_buffer_bytes]={0};

// callback used by A2DP to provide the sound data - usually len is 128 2 channel int16 frames
int32_t get_sound_data(Frame* data, int32_t len) {
  size_t req_bytes = min(max_buffer_bytes, len*2*sizeof(int16_t));
  // the microphone provides data in int32_t -> we read it into the buffer of int32_t data so *2
  //size_t result_bytes = i2s.readBytes((uint8_t*)buffer, req_bytes*2);
  size_t result_bytes = kit.readBytes((uint8_t*)buffer, req_bytes*2);
  // we have data only in 1 channel but we want to fill both
  return bothChannels.convert((uint8_t*)buffer, result_bytes);
}


// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  auto cfg = kit.defaultConfig(RX_MODE);
  cfg.sd_active = false;
  cfg.input_device = AUDIO_HAL_ADC_INPUT_LINE1; /* DIFFERENCE, ALL wie LINE1 mic input to adc channel 1  LINE2 does not work*/
  cfg.bits_per_sample = 16;
  cfg.channels = 2;
  cfg.sample_rate = 44100;
  cfg.is_master = true;
  kit.begin(cfg);
  kit.setVolume(100); //max?


  // start the bluetooth
  Serial.println("starting A2DP...");
  a2dp_source.start("BT-OEH1", get_sound_data);  
}

// Arduino loop - repeated processing 
void loop() {
}

**Comment by pschatzmann** [comment]
At first sight, the only thing which is not good is the callback: I2S for the Audiokit is already returning the data as needed with 2 channels, so no additinal buffer and conversion should be done: so a 
```
int32_t get_sound_data(Frame* data, int32_t len) {
  return kit.readBytes((uint8_t*)data, len);
}
```
should be sufficient...

**Comment by monopol40** [comment]
Many thanks for your hint. I changed get_sound_data according to your hint and indeed I hear some noise, but only that,  on left and right bt headphone, if I press my finger on left or right kitboard microphone. Speaking to the kitboard microphone produces rustling noice in both bt channels headphone. 
Are there possible some parameters to change to produce normal speech voice? Your sinus example with bt produces clear sinus sound on the same bt headphone.
Many thanks in advance!

**Comment by pschatzmann** [comment, marked as answer]
My mistake. The size is in frames and not bytes: see [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/basic-audiokit-a2dp/basic-audiokit-a2dp.ino)

**Reply by monopol40** [reply]
Many thanks, with size in frames it's now working!!!!
